cartoframes/data/utils.py

```diff
@@ -107,31 +121,109 @@ def wrapper(*args):
 
 
 @_encode_decode_decorator
-def decode_geometry(ewkb):
-    """Decode encoded wkb into a shapely geometry"""
-    # it's already a shapely object
-    if hasattr(ewkb, 'geom_type'):
-        return ewkb
-
+def decode_geometry(geom, enc_type):
+    """Decode any geometry into a shapely geometry."""
+    if geom:
+        func = {
+            ENC_SHAPELY: lambda: geom,
+            ENC_WKB: lambda: _load_wkb(geom),
+            ENC_WKB_HEX: lambda: _load_wkb_hex(geom),
+            ENC_WKB_BHEX: lambda: _load_wkb_bhex(geom),
+            ENC_WKT: lambda: _load_wkt(geom),
+            ENC_EWKT: lambda: _load_ewkt(geom)
+        }.get(enc_type)
+        return func() if func else geom
+    return ''
+
+
+def detect_encoding_type(input_geom):
+    """
+    Detect geometry encoding type:
+    - ENC_WKB: b'\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00H\x93@\x00\x00\x00\x00\x00\x9d\xb6@'
+    - ENC_EWKB: b'\x01\x01\x00\x00 \xe6\x10\x00\x00\x00\x00\x00\x00\x00H\x93@\x00\x00\x00\x00\x00\x9d\xb6@'
+    - ENC_WKB_HEX: '0101000000000000000048934000000000009DB640'
+    - ENC_EWKB_HEX: '0101000020E6100000000000000048934000000000009DB640'
+    - ENC_WKB_BHEX: b'0101000000000000000048934000000000009DB640'
+    - ENC_EWKB_BHEX: b'0101000020E6100000000000000048934000000000009DB640'
+    - ENC_WKT: 'POINT (1234 5789)'
+    - ENC_EWKT: 'SRID=4326;POINT (1234 5789)'
+    """
     from shapely import wkb
-    from shapely import wkt
-    if ewkb:
-        try:
-            return wkb.loads(ba.unhexlify(ewkb))
-        except Exception:
-            try:
-                return wkb.loads(ba.unhexlify(ewkb), hex=True)
-            except Exception:
+    from shapely.geometry.base import BaseGeometry
+
+    if isinstance(input_geom, BaseGeometry):
+        return ENC_SHAPELY
+
+    if isinstance(input_geom, str):
+        if re.match(r'^[0-9a-fA-F]+$', input_geom):
+            return ENC_WKB_HEX
+        else:
+            result = re.match(r'^SRID=\d+;(.*)$', input_geom)
+            if result:
+                geom = result.group(1)
+                if geom != '':
+                    return ENC_EWKT
+            elif input_geom != '':
+                # This is required because in P27 bytes = str
                 try:
-                    return wkb.loads(ewkb, hex=True)
+                    wkb.loads(input_geom)
+                    return ENC_WKB
                 except Exception:
-                    try:
-                        return wkb.loads(ewkb)
-                    except Exception:
-                        try:
-                            return wkt.loads(ewkb)
-                        except Exception:
-                            pass
+                    return ENC_WKT
+
+    if isinstance(input_geom, bytes):
+        try:
+            ba.unhexlify(input_geom)
+            return ENC_WKB_BHEX
+        except Exception:
+            return ENC_WKB
```

Comment:
This method looks a bit cryptic to me. A re-write proposal (I might be missing some case, but you get the idea):

```suggestion
    if isinstance(input_geom, BaseGeometry):
        return ENC_SHAPELY

    if isinstance(input_geom, str):
        if _is_hex(input_geom):
            return ENC_WKB_HEX
        else:
           srid, geom = _extract_srid(input_geom):
            if srid and geom:
                return ENC_EWKT
            elif geom:
                # This is required because in P27 bytes = str
                try:
                    wkb.loads(geom)
                    return ENC_WKB
                except Exception:
                    return ENC_WKT

    if isinstance(input_geom, bytes):
        try:
            ba.unhexlify(input_geom)
            return ENC_WKB_BHEX
        except Exception:
            return ENC_WKB
```